# Descritiva

In [96]:
import pandas as pd
import numpy as np

In [97]:
#importando a base de dados e atribuindo a variável df
df = pd.read_csv('diabetes.csv', sep = ',')
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [98]:
#Encontrando o maior valor da coluna Glucose
maxs = df["Glucose"].max()
maxs

199

In [99]:
#Encontrando o menor valor da coluna Glucose
mins = df["Glucose"].min()
mins

0

In [100]:
#Descobrir quais das colunas do dataset possuem valores nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [101]:
#Identificando se na base exixtem valores NULOS
df.isna()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin    BMI  \
0          False    False          False          False    False  False   
1          False    False          False          False    False  False   
2          False    False          False          False    False  False   
3          False    False          False          False    False  False   
4          False    False          False          False    False  False   
..           ...      ...            ...            ...      ...    ...   
763        False    False          False          False    False  False   
764        False    False          False          False    False  False   
765        False    False          False          False    False  False   
766        False    False          False          False    False  False   
767        False    False          False          False    False  False   

     DiabetesPedigreeFunction    Age  Outcome  
0                       False  False    False  
1                       False  False    False  
2                       False  False    False  
3                       False  False    False  
4                       False  False    False  
..                        ...    ...      ...  
763                     False  False    False  
764                     False  False    False  
765                     False  False    False  
766                     False  False    False  
767                     False  False    False  

[768 rows x 9 columns]

In [102]:
#Identificando valores 0 na base
#df.isnull().sum() nao funcionou
#df[df['Age'] <= 0].value_counts()

#df.loc[df[]==0]

In [103]:
#Linhas onde age é igual a 30
#Como informar o total?
df.loc[df['Age'] == 30]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
5              5      116             74              0        0  25.6   
10             4      110             92              0        0  37.6   
52             5       88             66             21       23  24.4   
126            3      120             70             30      135  42.9   
132            3      170             64             37      225  34.5   
256            3      111             56             39        0  30.1   
300            0      167              0              0        0  32.3   
309            2      124             68             28      205  32.9   
316            3       99             80             11       64  19.3   
326            1      122             64             32      156  35.1   
351            4      137             84              0        0  31.2   
364            4      147             74             25      293  34.9   
365            5       99             54             28       83  34.0   
444            4      117             62             12        0  29.7   
496            5      110             68              0        0  26.0   
556            1       97             70             40        0  38.1   
602            1      124             74             36        0  27.8   
616            6      117             96              0        0  28.7   
645            2      157             74             35      440  39.4   
706           10      115              0              0        0   0.0   
765            5      121             72             23      112  26.2   

     DiabetesPedigreeFunction  Age  Outcome  
5                       0.201   30        0  
10                      0.191   30        0  
52                      0.342   30        0  
126                     0.452   30        0  
132                     0.356   30        1  
256                     0.557   30        0  
300                     0.839   30        1  
309                     0.875   30        1  
316                     0.284   30        0  
326                     0.692   30        1  
351                     0.252   30        0  
364                     0.385   30        0  
365                     0.499   30        0  
444                     0.380   30        1  
496                     0.292   30        0  
556                     0.218   30        0  
602                     0.100   30        0  
616                     0.157   30        0  
645                     0.134   30        0  
706                     0.261   30        1  
765                     0.245   30        0

In [104]:
#Linhas onde age é maior que 30
df.loc[df['Age'] > 30]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
4              0      137             40             35      168  43.1   
8              2      197             70             45      543  30.5   
..           ...      ...            ...            ...      ...   ...   
759            6      190             92              0        0  35.5   
761            9      170             74             31        0  44.0   
762            9       89             62              0        0  22.5   
763           10      101             76             48      180  32.9   
766            1      126             60              0        0  30.1   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
4                       2.288   33        1  
8                       0.158   53        1  
..                        ...  ...      ...  
759                     0.278   66        1  
761                     0.403   43        1  
762                     0.142   33        0  
763                     0.171   63        0  
766                     0.349   47        1  

[351 rows x 9 columns]

In [105]:
#desvio padrao aproximado da coluna "BloodPressure""

df.std()

Pregnancies                   3.369578
Glucose                      31.972618
BloodPressure                19.355807
SkinThickness                15.952218
Insulin                     115.244002
BMI                           7.884160
DiabetesPedigreeFunction      0.331329
Age                          11.760232
Outcome                       0.476951
dtype: float64

In [106]:
#Qual a proporcao aproximada de elementos da classe 0 e da classe 1?
#funcao normalize traz a porcentagem de ocorrencia, como padrão ele vem false, podemos alterar para true
df['Outcome'].value_counts(dropna=False, normalize=True)*100

0    65.104167
1    34.895833
Name: Outcome, dtype: float64

In [109]:
#trazer valores 0 da coluna Glucose
df[df['Glucose'].isin([0])]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
75             1        0             48             20        0  24.7   
182            1        0             74             20       23  27.7   
342            1        0             68             35        0  32.0   
349            5        0             80             32        0  41.0   
502            6        0             68             41        0  39.0   

     DiabetesPedigreeFunction  Age  Outcome  
75                      0.140   22        0  
182                     0.299   21        0  
342                     0.389   22        0  
349                     0.346   37        1  
502                     0.727   41        1

In [ ]:
#19
#substituir os valores 0 por Nan (pendente)
df['Glucose'] = df['Glucose'].replace(0, np.NaN)

In [111]:
df[df['Glucose'].isnull()]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
75             1      NaN             48             20        0  24.7   
182            1      NaN             74             20       23  27.7   
342            1      NaN             68             35        0  32.0   
349            5      NaN             80             32        0  41.0   
502            6      NaN             68             41        0  39.0   

     DiabetesPedigreeFunction  Age  Outcome  
75                      0.140   22        0  
182                     0.299   21        0  
342                     0.389   22        0  
349                     0.346   37        1  
502                     0.727   41        1

# Questão 20

Crie um código para preencher os valores faltantes de glucose utilizando como valor, a media dos valores de 'Glucose' separada por grupo de idade, opu seja, se a pessoa sem dado de "Glucose" tiver 21 anos utilize media de Glucose de 21 anos para preencher o valor faltante.

Apos o preenchimento dos dados faltantes, selecione todas as alternativas abaixo onde temos um par de "Idade"(Age)
e "Glucose Media" utilizados para o preenchimento de ao menos um valor faltante.

Por exemplo, se no conjunto de daos tivermos um paciente com 33 anos sem dados de Glucose, selecione a opcao onde temos "idade: 33" e onde a "Glucose media" é igual a "Glucose média"para as pessoas de 33 anos.

Obs: Valores foram truncados, desconsiderando as casas decimais.

Resposta:
Idade: 21 Glucose média: 110
Idade: 22 Glucose média: 111
Idade: 37 Glucose média: 137
Idade: 41 Glucose média: 135

In [115]:
#20
#Printar a idade media
dados_1 = df[df['Glucose'].isnull()]
idade_2 = dados_1['Age']
for x in idade_2:
    dados_aux = df[df['Age'] == x]
    z = dados_aux['Glucose'].mean()
    print(x,' ',z)

22   111.3
21   110.06451612903226
22   111.3
37   137.38888888888889
41   135.23809523809524


In [215]:
#substituir valores faltantes pela média
#codigo abaixo não funcionou para subistituir
#df['Glucose'] = df['Glucose'].replace(np.NaN, z)
#21
df['Glucose'] = df['Glucose'].replace(z, np.NaN)

In [130]:
#df.loc[182]

Pregnancies                  1.000
Glucose                        NaN
BloodPressure               74.000
SkinThickness               20.000
Insulin                     23.000
BMI                         27.700
DiabetesPedigreeFunction     0.299
Age                         21.000
Outcome                      0.000
Name: 182, dtype: float64

In [216]:
df[df['Glucose'].isnull()]

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

# Questão 22

Descarte os tratamentos de valores faltantes realizados na questão anterior, e realize a separação dos dados em um conjunto de treino e de teste (de forma estratificada) e selecione AS alternativas corretas.

Utilize os seguintes parâmetros:
-test_size= 0,4
-random_state= 0 
-stratify = y (variável target)

R: Temos 299 targets "0"no conjunto de TREINO
   Temos 107 targets "1" no conjunto de TESTE
   Temos 460 observações no conjunto de TREINO
   Temos 308 observações no conjunto de TESTE
    

In [217]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

import matplotlib.pyplot as plt
plt.style.use('ggplot')

In [218]:
df= pd.read_csv('diabetes.csv')

In [219]:
df['Glucose'] = df['Glucose'].replace(0, np.NaN)

In [220]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6    148.0             72             35        0  33.6   
1            1     85.0             66             29        0  26.6   
2            8    183.0             64              0        0  23.3   
3            1     89.0             66             23       94  28.1   
4            0    137.0             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [221]:
df.Outcome.value_counts(normalize = True)*100

0    65.104167
1    34.895833
Name: Outcome, dtype: float64

In [222]:
#y é a variável target, e X é a variavel teste
y = df['Outcome']
X = df.drop('Outcome', axis = 1)

In [149]:
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0             72             35        0  33.6   
1              1     85.0             66             29        0  26.6   
2              8    183.0             64              0        0  23.3   
3              1     89.0             66             23       94  28.1   
4              0    137.0             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10    101.0             76             48      180  32.9   
764            2    122.0             70             27        0  36.8   
765            5    121.0             72             23      112  26.2   
766            1    126.0             60              0        0  30.1   
767            1     93.0             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [240]:
#stratify indica que quero a mesma proporçao de y da linha 148
X_train, X_test,y_train,y_test = train_test_split(X,y,
                                                  test_size=0.4,
                                                  random_state=0,
                                                  stratify=y)

In [224]:
scaler = MinMaxScaler()

In [241]:
scaler.fit(X_train)

MinMaxScaler()

In [242]:
#escalonando todos os dados de treino e teste em 0 e 1
X_train[X_train.columns] = scaler.transform(X_train)
X_test[X_test.columns] = scaler.transform(X_test)

<ipython-input-242-aafc9d817842>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train[X_train.columns] = scaler.transform(X_train)
/Users/thaismarcondes/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1738: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value[:, i].tolist(), pi)
<ipython-input-242-aafc9d817842>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value in

In [243]:
X_train[X_train.columns]

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
741     0.176471  0.374194       0.360656       0.317460  0.111111  0.459016   
619     0.000000  0.483871       0.000000       0.000000  0.000000  0.482861   
317     0.176471  0.890323       0.606557       0.000000  0.000000  0.454545   
599     0.058824  0.419355       0.311475       0.285714  0.141844  0.344262   
512     0.529412  0.303226       0.557377       0.000000  0.000000  0.360656   
..           ...       ...            ...            ...       ...       ...   
456     0.058824  0.587097       0.442623       0.000000  0.000000  0.397914   
578     0.588235  0.574194       0.557377       0.000000  0.000000  0.402385   
398     0.176471  0.245161       0.573770       0.000000  0.000000  0.314456   
464     0.588235  0.458065       0.803279       0.000000  0.000000  0.357675   
363     0.235294  0.658065       0.639344       0.000000  0.000000  0.573770   

     DiabetesPedigreeFunction       Age  
741                  0.137489  0.083333  
619                  0.026900  0.050000  
317                  0.114005  0.133333  
599                  0.140478  0.083333  
512                  0.052092  0.616667  
..                        ...       ...  
456                  0.260034  0.683333  
578                  0.071307  0.250000  
398                  0.132792  0.066667  
464                  0.403074  0.216667  
363                  0.188728  0.766667  

[460 rows x 8 columns]

In [244]:
X_test[X_test.columns]

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
691     0.764706  0.735484       0.934426       0.000000  0.000000  0.630402   
94      0.117647  0.632258       0.672131       0.285714  0.075650  0.368107   
415     0.176471  0.832258       0.688525       0.523810  0.560284  0.532042   
470     0.058824  0.645161       0.672131       0.634921  0.000000  0.615499   
169     0.176471  0.432258       0.737705       0.190476  0.092199  0.423249   
..           ...       ...            ...            ...       ...       ...   
743     0.529412  0.619355       0.770492       0.000000  0.000000  0.487332   
200     0.000000  0.445161       0.655738       0.253968  0.000000  0.461997   
232     0.058824  0.225806       0.655738       0.396825  0.043735  0.378539   
418     0.058824  0.251613       0.557377       0.000000  0.000000  0.271237   
181     0.000000  0.483871       0.524590       0.285714  0.108747  0.520119   

     DiabetesPedigreeFunction       Age  
691                  0.076430  0.383333  
94                   0.291631  0.000000  
415                  0.076857  0.016667  
470                  0.225875  0.116667  
169                  0.178053  0.133333  
..                        ...       ...  
743                  0.280102  0.400000  
200                  0.339880  0.000000  
232                  0.215628  0.016667  
418                  0.233134  0.100000  
181                  0.276260  0.033333  

[308 rows x 8 columns]

In [245]:
y_train.value_counts()

0    299
1    161
Name: Outcome, dtype: int64

In [246]:
y_test.value_counts()

0    201
1    107
Name: Outcome, dtype: int64

In [247]:
#dados nulos da coluna Glucose na base de teste
X_test[X_test['Glucose'].isnull()]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin       BMI  \
75      0.058824      NaN       0.393443       0.317460      0.0  0.368107   
342     0.058824      NaN       0.557377       0.555556      0.0  0.476900   

     DiabetesPedigreeFunction       Age  
75                   0.026473  0.016667  
342                  0.132792  0.016667

# Questão 23

Por simplicidade, realize o tratamento de missing values da coluna "Glucose"utilizando o "SimpleImputer"do SKLearn, em seguida responda:

Qual foi o valor médio aproximado de "Glucose"utilizado pelo "SimpleImputer"para realizar a inputação de valores faltantes?

Obs1: Esta função não permite o cálculo de médias separadas por grupo de idade, sendo assim, a imputação dos valores faltantes será realizada puramente utilizando-se média geral da coluna

Obs2: Lembre-se de substituir os valores de Glucose iguais a zero por "np.NaN"antes de realizar a imputação, ou altere o parâmetro 'missing_values'

R: 121

In [248]:
from sklearn.impute import SimpleImputer

imp = SimpleImputer(missing_values=np.nan, strategy='mean')
df['Glucose'] = imp.fit_transform(df[['Glucose']])
df[['Glucose','Glucose']].head(10)


Glucose  Glucose
0    148.0    148.0
1     85.0     85.0
2    183.0    183.0
3     89.0     89.0
4    137.0    137.0
5    116.0    116.0
6     78.0     78.0
7    115.0    115.0
8    197.0    197.0
9    125.0    125.0

In [249]:
#resposta 121
df.loc[182]

Pregnancies                   1.000000
Glucose                     121.686763
BloodPressure                74.000000
SkinThickness                20.000000
Insulin                      23.000000
BMI                          27.700000
DiabetesPedigreeFunction      0.299000
Age                          21.000000
Outcome                       0.000000
Name: 182, dtype: float64

# Questão 24

Utilizando o mesmo SimpleInputer criado na questão anterior, utilize o método "transform"nos dados de teste para realizar a inputação dos valores faltantes do conjunto de teste e responda:

Qual foi o valor de Glucose aproximado utilizado pelo "SimpleInputer" para realizar a imputação de valores faltantes no conjunto de teste?

R: 121

In [250]:
imp = SimpleImputer(missing_values=np.nan, strategy='mean')
X_test['Glucose'] = imp.fit_transform(X_test[['Glucose']])
X_test[['Glucose','Glucose']].head(10)

<ipython-input-250-df238507e318>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test['Glucose'] = imp.fit_transform(X_test[['Glucose']])


Glucose   Glucose
691  0.735484  0.735484
94   0.632258  0.632258
415  0.832258  0.832258
470  0.645161  0.645161
169  0.432258  0.432258
151  0.451613  0.451613
125  0.283871  0.283871
654  0.400000  0.400000
296  0.658065  0.658065
765  0.496774  0.496774

In [251]:
#Resposta 0,50 incorreta, entender pq deste resultado
X_test.loc[75]

Pregnancies                 0.058824
Glucose                     0.505587
BloodPressure               0.393443
SkinThickness               0.317460
Insulin                     0.000000
BMI                         0.368107
DiabetesPedigreeFunction    0.026473
Age                         0.016667
Name: 75, dtype: float64

# Questao 25

Realize a reescala dos atributos do conjunto de treino utilizando o "MinMaxScaler"e responda:

Quais valores foram obtidos ao se observar os valores do atributo "data_max"do objeto "MinMaxScaler"?

R: [17.199.122.63.846.67.1 2.42 81]

In [256]:
MinMaxScaler()
print(scaler.fit(X_train))
print(scaler.data_max_)

MinMaxScaler()
[1. 1. 1. 1. 1. 1. 1. 1.]


# Questão 26

Utilizando o "MinMaxScaler"ajustado no conjunto de treino na questão anterior, realize a normalização dos dados no conjunto de treino e de teste e responda:

Qual o valor médio aproximado da coluna "Glucose"após a aplicação da normalização dos dados no conjunto de TREINO?

R: 0.498

# Questão 27

Ainda com base na normalização dos dados realizada na questão anterior, responda:

Qual o número médio aproximado da coluna 'Glucose'após a aplicação da normalização dos dados no conjunto de TESTE?

R: 0.505

# Questão 28

Crie um modelo KNN utilizando 5 vizinhos, realize o treinamento desde modelo nos dados de treino e responda:

Qual foi a acurácia aproximada do modelo no conjunto de TREINO?

R: 0.82

# Questão 29

Ainda com base no modelo criado na questão anterior, qual foi o valor aproximado de acurácia obtido no conjunto de TESTE?

R: 0.72

# Questão 30

Defina um novo modelo de KNN utilizando 5 vizinhos, altere o parâmetro "weights = 'distance', realize seu treinamento no conjunto de treino e responda:

Qual foi a acurácia aproximada do modelo no conjunto de TREINO?

R: 1.0

# Questão 31

Com base no mesmo modelo da questão anterior, qual foi o valor aproximado de acurácia obtido no conjunto de TESTE?

R: 0.72